faculty_mill/publish.py

```diff
@@ -0,0 +1,94 @@
+import os
+from pathlib import Path
+from typing import Optional
+from uuid import UUID
+from time import sleep
+import click
+import sherlockml
+import sml.auth
+
+
+def publish(
+    report_name: str,
+    path: Path,
+    show_code: bool = False,
+    report_id: Optional[UUID] = None,
+    project_id: Optional[UUID] = None,
+    user_id: Optional[UUID] = None,
+):
+    """Publish a notebook as a report.
+
+    Parameters
+    ----------
+    report_name : str
+        The name of the report
+    path : Path
+        The path of the notebook.
+    report_id : Optional[UUID], optional
+        The report ID, if you want to publish it as a version of an existing
+        report (the default is None, in which case we search for an existing
+        report with the provided name)
+    project_id : Optional[UUID], optional
+        The project ID. Only needed if not invoking from within a project.
+    user_id : Optional[UUID], optional
+        The user ID. Only needed if not invoking from within a project.
+    show_code : bool, optional
+        Whether the code should be shown in the report or not (default False)
+    """
+
+    if project_id is None:
+        project_id = UUID(os.getenv("SHERLOCKML_PROJECT_ID"))
```

Comment:
I prefer `project_id  = project_id or UUID(os.getenv("SHERLOCKML_PROJECT_ID"))`

Reply:
this is why I had it initially, but in this case it's either a UUID or None, so it seems safe

Reply:
I think in the spirit of "Explicit is better than implicit", and being consistent so we don't get in the habit of using the shortcut in contexts where it does behave differently, the `is None` check is better.

Reply:
Given there's the `Optional[UUID]` type hint, I think either is acceptable, however I prefer the clarity that comes with the ` if _ is None:` statement.

I know that mypy is able to use the `if project_id is None` test to know that `project_id` is a `UUID` and not an `Optional[UUID]` on subsequent lines - I'm not sure if that is also True under @liamcoatman's proposed change.

Reply:
In fact, mypy can't use `if something` to disambiguate `None`'s, so that's probably a good reason to be explicit

Reply:
With regards to @srstevenson original concern testing whether the argument is falsey, we definitely don't want to allow a `project_id` argument value like `False` to be overridden from the environment. IMO `if _ is None` is definitely needed.